File: src/mod10_design_pattern/lesson14_iterator/java/CollectionImpl.java
package mod10_design_pattern.lesson14_iterator.java;

public class CollectionImpl<T> implements ICollection<T> {

    private T[] items;
    private int size;

    public CollectionImpl(T[] items) {
        this.items = items;
        this.size = items.length;
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<T>() {

            private int index = 0;

            @Override
            public boolean hasNext() {
                return index < size;
            }

            @Override
            public T next() {
                return items[index++];
            }
        };
    }
}

File: src/mod10_design_pattern/lesson14_iterator/kotlin/CollectionImpl.kt
package mod10_design_pattern.lesson14_iterator.kotlin

class CollectionImpl<T>(
    private val items: Array<T>
) : Collection<T> {

    private var size = items.size

    override fun iterator(): Iterator<T> {
        return object : Iterator<T> {

            var index = 0

            override fun hasNext() = index < size

            override fun next(): T {
                if (!hasNext()) throw NoSuchElementException()
                return items[index++]
            }
        }
    }
}